File: src/data/courses/dsa/binary-search-on-answer/rate-speed-problems/minimum-eating-speed/code/cpp/minimum_eating_speed.cpp
class Solution {
public:
    // Same as Koko Eating Bananas
    int minEatingSpeed(vector<int>& piles, int h) {
        int lo = 1, hi = *max_element(piles.begin(), piles.end());
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            long hours = 0;
            for (int p : piles) hours += (p + mid - 1) / mid;
            if (hours <= h) hi = mid; else lo = mid + 1;
        }
        return lo;
    }
};

File: src/data/courses/dsa/binary-search-on-answer/rate-speed-problems/minimum-eating-speed/code/java/MinimumEatingSpeed.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int lo=1,hi=0; for(int p:piles)hi=Math.max(hi,p);
        while(lo<hi){int mid=lo+(hi-lo)/2;long hrs=0;for(int p:piles)hrs+=(p+mid-1)/mid;if(hrs<=h)hi=mid;else lo=mid+1;}
        return lo;
    }
}